### ISSUE TITLE
[function]  - curved lines / zoom issues

### ISSUE THREAD

**Issue by porres** [porres/pd-else#641]

**Comment by porres** [comment]
@sebshader , you seem to be a tcl/tk genious, could you give me a hint on what the direction is to draw curved like shapes for the envelope generator [function] object?

**Comment by sebshader** [comment]
@porres there's a reason that curved patch cables are not a feature that's widely used in vanilla.. in tcl/tk without any extensions curved lines look a bit.. rough.
at any rate, the way to do it in regular tk (without extensions) is with the -smooth and -splinesteps options to the "create line" canvas command. This creates bezier curves.
https://www.tcl.tk/man/tcl8.4/TkCmd/canvas.htm#M144
for the curved cables in pd-next I used -splinesteps 36 but it might be fine to use fewer in ```[function]``` since the area is smaller. (something like 14 maybe?)

**Comment by porres** [comment]
what about things like [scope~]? where you can plot a wave signal with any shape?

I want to make curves that follow an exponential rule, like raising to the power f 4 and they need to look the same as the signal output they generate

**Comment by sebshader** [comment]
@porres I think that scope~ and the arrays in vanilla often just use a lot of line segments, which can look better. In that case you might have some custom drawing function in c and you linearly interpolate it with line segments. That would also work.
I think that's what the -splinesteps option is supposed to do as well though. (and it wouldn't have to be by hand). But it's possible you might have more luck mapping the points explicitly in c and then drawing them. 

so if you do it that way you'd have a function for whatever curve you want to generate and sample it at regular intervals, and then use the tk canvas ```create line``` to draw lines between the samples.

But I would just use the built in curves if I were you.
(the "bezier curve" option in the array properties uses the native tk bezier curves, so if you try that with only a few points in the array I think that's kind of how it would look probably)

**Comment by sebshader** [comment]
just toying with it right now it looks like fewer splinesteps might actually look better. I like ~12-14

**Comment by porres** [comment]
> I think that scope~ and the arrays in vanilla often just use a lot of line segments

yeah, that's what I am doing now... I need to do this because I am setting specific exponential values.
